### ISSUE TITLE
DDP: moving model to CPU and back to GPU breaks gradient synchronization

### ISSUE THREAD

**Issue by vlievin** [pytorch/pytorch#104336]
### 🐛 Describe the bug

Using `torch.nn.parallel.DistributedDataParallel`,  gradient synchronisation (in `torch.backward`) breaks when moving a DDP-wrapped model back to CPU and back again to GPU. After moving the model between CUDA -> CPU -> CUDA, running subsequent `torch.backward()` won't trigger gradient reduction (averaging gradients between workers). This causes the model to be effectively trained on a single device (without throwing any warning/error).

**Use case**
Moving a model temporarily to CPU is useful when GPU resources need to be temporarily allocated for another task (e.g., building a faiss index).

**Expected behaviour**
I think `DDP` should support moving models back and forth to device or `DDP` should throw a warning when calling `DistributedDataParallel.cpu()`.

**Code to reproduce the error:**
```python
import os

import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn as nn
import torch.optim as optim
from torch.nn.parallel import DistributedDataParallel as DDP


class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = torch.nn.Linear(10, 10)
        self.relu = torch.nn.ReLU()
        self.net2 = torch.nn.Linear(10, 5)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = self.relu(self.net1(x))
        return self.net2(x)


def ddp_check_grads_synchronized(move_model_to_cpu_and_back: bool) -> None:
    world_size = torch.cuda.device_count()
    mp.spawn(
        _ddp_check_grads_synchronized,
        args=(
            world_size,
            move_model_to_cpu_and_back,
        ),
        nprocs=world_size,
        join=True,
    )


def _ddp_check_grads_synchronized(rank: int, world_size: int, move_model_to_cpu_and_back: bool) -> None:
    print(f"Running basic DDP example on rank {rank}.")
    _ddp_setup(rank, world_size)

    # create model and move it to GPU with id rank
    model = ToyModel().to(rank)
    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(model.parameters(), lr=0.001)
    ddp_model = DDP(model, device_ids=[rank])

    if move_model_to_cpu_and_back:
        # move the model to CPU and back to GPU
        # this is to test that the gradients are synchronized even if the model is moved to CPU
        ddp_model.cpu()
        ddp_model.cuda(rank)

    # create data
    x = torch.randn(20, 10)
    labels = torch.randn(20, 5)

    # forward pass
    x = x.to(rank)
    labels = labels.to(rank)
    outputs = ddp_model(x)

    # backward pass
    optimizer.zero_grad()
    loss = loss_fn(outputs, labels)
    loss.backward()

    _check_gradients_between_workers(
        world_size,
        move_model_to_cpu_and_back,
        ddp_model,
        verbose=rank == 0,
    )
    _ddp_cleanup()


def _check_gradients_between_workers(
    world_size: int,
    move_model_to_cpu_and_back: bool,
    model: nn.Module,
    verbose: bool = False,
) -> None:
    """Gather gradients from all processes and check that they are equal."""
    for k, v in model.named_parameters():
        if v.grad is not None:
            grad_list = [torch.zeros_like(v.grad) for _ in range(world_size)]
            dist.all_gather(grad_list, v.grad, async_op=False)
            for grad in grad_list:
                if not torch.allclose(v.grad, grad):
                    if verbose:
                        print(
                            {
                                "grad": v.grad.mean(),
                                "other_grad": grad.mean(),
                            }
                        )
                    raise RuntimeError(
                        f"move_model_to_cpu_and_back={move_model_to_cpu_and_back}. "
                        f"Gradients are not equal across processes (p={k})"
                    )


def _ddp_setup(rank: int, world_size: int) -> None:
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12367"
    dist.init_process_group("gloo", rank=rank, world_size=world_size)


def _ddp_cleanup() -> None:
    dist.destroy_process_group()


if __name__ == "__main__":
    ddp_check_grads_synchronized(move_model_to_cpu_and_back=False)
    print("\nmove_model_to_cpu_and_back=False: SUCCESS!\n")

    ddp_check_grads_synchronized(move_model_to_cpu_and_back=True)
    print("\nmove_model_to_cpu_and_back=True: SUCCESS!\n")

```

### Versions

PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.28

Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-4.19.0-24-cloud-amd64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB

Nvidia driver version: 530.30.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 48 bits virtual
CPU(s):              24
On-line CPU(s) list: 0-23
Thread(s) per core:  2
Core(s) per socket:  12
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:            7
CPU MHz:             2200.216
BogoMIPS:            4400.43
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            39424K
NUMA node0 CPU(s):   0-23
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] torch==2.0.1
[pip3] triton==2.0.0
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by kumpera** [comment]
The problem of calling `cpu` or `cuda` is that modifies the backing tensors of the model's Parameters.
DDP register autograd hooks to the original tensors and it won't see the new ones.

There are two ways to address your issue.

1) Rewrap the model with DDP around moving between devices.

For that you can use this code snippet:
```python
    if move_model_to_cpu_and_back:
        # move the model to CPU and back to GPU
        # this is to test that the gradients are synchronized even if the model is moved to CPU
        del ddp_model
        model.cpu()
        model.to(rank)
        ddp_model = DDP(model, device_ids=[rank])
```

2) play around with tensors storage and restore them around your model. This is very tricky and will likely cause subtle bugs if you're not careful enough.

It would be something like this:

```python
from torch.distributed.utils import _alloc_storage, _free_storage

with torch.no_grad():
    cpu_data = {}
    print("offloading")
    for name, param in ddp_model.named_parameters():
        cpu_data[name] = (param.data.cpu(), param.data.size())
        _free_storage(param.data)

    print("reloading")
    for name, param in ddp_model.named_parameters():
        cpu_tensor, size = cpu_data[name]
        _alloc_storage(param.data, size)
        param.data.copy_(cpu_tensor)
```

**Comment by satoshi-iwata** [comment]
> DDP register autograd hooks to the original tensors and it won't see the new ones.

Hi, could anyone tell me where these hooks are registered?

I checked some hooks registered to a DDP model and parameters but couldn't find them.
I've checked
- DDP model's hooks such as `_backward_hooks` and `_accum_grad_hooks`
- parameter's hooks `_backward_hooks` and `_post_accumulate_grad_hooks`
